src/Stratis.Bitcoin/Base/AsyncWork/AsyncProvider.cs

```diff
@@ -0,0 +1,212 @@
+        /// <param name="state">The <see cref="IAsyncDelegateDequeuer"/> that's run by the delegateTask</param>
+        private void onAsyncDelegateUnhandledException(Task task, object state)
+        {
+            AsyncTaskInfo delegateInfo;
+            lock (this.lockAsyncDelegates)
+            {
+                if (this.asyncDelegateWorkers.TryGetValue((IAsyncDelegateDequeuer)state, out delegateInfo))
+                {
+                    IAsyncTaskInfoSetter infoSetter;
+                    infoSetter = (IAsyncTaskInfoSetter)delegateInfo;
+                    infoSetter.Exception = task.Exception.GetBaseException();
+                    infoSetter.Status = task.Status;
+                }
+                else
+                {
+                    // Should never happen.
+                    this.logger.LogError("Cannot find the AsyncDelegateInfo related to the faulted task with Id {0}", task.Id);
+                    return;
+                }
+
+                this.logger.LogError(task.Exception.GetBaseException(), "Unhandled exception for async delegate worker {0}", delegateInfo.FriendlyName);
+            }
+        }
+
+        /// <summary>
+        ///  This method is called when delegateTask completed or was canceled.
+        ///  It removes the delegate worker from the dictionary.
+        /// </summary>
+        /// <param name="task">The delegate task.</param>
+        /// <param name="state">The IAsyncDelegate that's run by the delegateTask</param>
+        private void onAsyncDelegateCompleted(Task task, object state)
+        {
+            bool removed;
+            lock (this.lockAsyncDelegates)
+            {
+                removed = this.asyncDelegateWorkers.Remove((IAsyncDelegateDequeuer)state);
+            }
+
+            if (removed)
+            {
+                this.logger.LogTrace("Async delegate worker task removed. Id: {0}", task.Id);
+            }
+            else
+            {
+                // Should never happen.
+                this.logger.LogError("Cannot find the async delegate worker task with Id {0}", task.Id);
+            }
+        }
+
+        /// <inheritdoc />
+        public IAsyncLoop CreateAndRunAsyncLoop<T>(string name, Func<CancellationToken, Task> loop, CancellationToken cancellation, TimeSpan? repeatEvery = null, TimeSpan? startAfter = null)
+        {
+            Guard.NotEmpty(name, nameof(name));
+            Guard.NotNull(loop, nameof(loop));
+            Guard.NotNull(cancellation, nameof(cancellation));
+
+            // instantiate the loop
+            IAsyncLoop loopInstance = new AsyncLoop(name, this.logger, loop);
+
+            Task loopTask;
+            lock (this.lockAsyncLoops)
+            {
+                this.asyncLoops.Add(loopInstance, new AsyncTaskInfo(name));
+            }
+
+            loopTask = loopInstance.Run(cancellation, repeatEvery ?? TimeSpan.FromMilliseconds(DefaultLoopRepeatInterval), startAfter).RunningTask;
```

Comment:
Ther emight be a race condition if the run falts beoe you assign the ContinueWith (or mayeb not I am not sure how Task behaves in late binding

Reply:
no you shouldn't get any exception even if it ends before because there isn't any wait (that would trigger any task exception it has).
Attaching the continue with later is not an issue because if the task is in faulty state, will trigger the continue with straight at that point
I can't even do differently in this case, without changing how AsyncLoop works, because RunningTask is set by the Run method.

I will add a test to test this condition

Reply:
yes good idea (maybe you'll have to attach the faulty first as otherwise it will thro win the none faulty continuation)

Reply:
actually this is what I do:
first one uses TaskContinuationOptions.OnlyOnFaulted so it continue ONLY when the task is faulty
the second uses TaskContinuationOptions.NotOnFaulted so it will continue when the task ends with any state EXCEPT faulted